File: lua/mutate4lua/legacy/report.lua
local report = {}
function report.scan(relative_file, sites, changed_scope_ids)
  local lines = {
    string.format("Scan: %d mutation sites in %s", #sites, relative_file),
  }
  for _, site in ipairs(sites) do
    local marker = changed_scope_ids and changed_scope_ids[site.scope_id] and "* " or "  "
    lines[#lines + 1] = string.format("%s%s:%d %s", marker, relative_file, site.line, site.description)
  end
  if changed_scope_ids and next(changed_scope_ids) ~= nil then
    lines[#lines + 1] = "* indicates a scope that differs from the embedded manifest."
  end
  return table.concat(lines, "\n") .. "\n"
end
function report.diagnostics(selection_result, mutation_warning)
  local lines = {
    "Total mutation sites: " .. selection_result.total_mutation_sites,
    "Covered mutation sites: " .. #selection_result.covered,
    "Uncovered mutation sites: " .. #selection_result.uncovered,
    "Changed mutation sites: " .. selection_result.changed_mutation_sites,
    "Manifest exists: " .. tostring(selection_result.manifest_exists),
    "Project hash changed: " .. tostring(selection_result.project_hash_changed),
    "Differential surface area: " .. selection_result.differential_surface_area,
    "Manifest-violating surface area: " .. selection_result.manifest_violating_surface_area,
  }
  if #selection_result.covered == 0 then
    lines[#lines + 1] = "No mutations need testing."
  elseif #selection_result.covered > mutation_warning then
    lines[#lines + 1] = string.format("WARNING: Found %d mutations. Consider splitting this module.", #selection_result.covered)
  end
  return table.concat(lines, "\n") .. "\n"
end
function report.run(relative_file, baseline, diagnostics, uncovered, results)
  local diagnostics_text = diagnostics:gsub("\n$", "")
  local lines = {
    string.format("Baseline tests passed in %d ms.", baseline.duration_ms),
    diagnostics_text,
  }
  for _, site in ipairs(uncovered) do
    lines[#lines + 1] = string.format("UNCOVERED %s:%d %s", relative_file, site.line, site.description)
  end
  local killed = 0
  for _, result in ipairs(results) do
    if result.killed then
      killed = killed + 1
    end
    lines[#lines + 1] = string.format("%s %s:%d %s (%d ms)", result.killed and "KILLED" or "SURVIVED", relative_file, result.line, result.description, result.duration_ms)
    if result.timed_out then
      lines[#lines + 1] = "  timed out"
    end
  end
  lines[#lines + 1] = string.format("Coverage: %d uncovered sites skipped.", #uncovered)
  lines[#lines + 1] = string.format("Summary: %d killed, %d survived, %d total.", killed, #results - killed, #results)
  return table.concat(lines, "\n") .. "\n"
end
return report

File: internal/report/report.go
package report

import (
	"fmt"
	"strings"

	"github.com/billyq/mutate4lua/internal/analysis"
	mutruntime "github.com/billyq/mutate4lua/internal/runtime"
)

type DiagnosticsInput struct {
	TotalMutationSites           int
	CoveredCount                 int
	UncoveredCount               int
	ChangedMutationSites         int
	ManifestExists               bool
	ProjectHashChanged           bool
	DifferentialSurfaceArea      int
	ManifestViolatingSurfaceArea int
}

func ScanReport(relativeFile string, sites []analysis.MutationSite, changedScopeIDs map[string]bool) string {
	lines := []string{fmt.Sprintf("Scan: %d mutation sites in %s", len(sites), relativeFile)}
	for _, site := range sites {
		marker := "  "
		if changedScopeIDs != nil && changedScopeIDs[site.ScopeID] {
			marker = "* "
		}
		lines = append(lines, fmt.Sprintf("%s%s:%d %s", marker, relativeFile, site.Line, site.Description))
	}
	if len(changedScopeIDs) > 0 {
		lines = append(lines, "* indicates a scope that differs from the embedded manifest.")
	}
	return strings.Join(lines, "\n") + "\n"
}

func DiagnosticsReport(input DiagnosticsInput, mutationWarning int) string {
	lines := []string{
		fmt.Sprintf("Total mutation sites: %d", input.TotalMutationSites),
		fmt.Sprintf("Covered mutation sites: %d", input.CoveredCount),
		fmt.Sprintf("Uncovered mutation sites: %d", input.UncoveredCount),
		fmt.Sprintf("Changed mutation sites: %d", input.ChangedMutationSites),
		fmt.Sprintf("Manifest exists: %t", input.ManifestExists),
		fmt.Sprintf("Project hash changed: %t", input.ProjectHashChanged),
		fmt.Sprintf("Differential surface area: %d", input.DifferentialSurfaceArea),
		fmt.Sprintf("Manifest-violating surface area: %d", input.ManifestViolatingSurfaceArea),
	}
	if input.CoveredCount == 0 {
		lines = append(lines, "No mutations need testing.")
	} else if input.CoveredCount > mutationWarning {
		lines = append(lines, fmt.Sprintf("WARNING: Found %d mutations. Consider splitting this module.", input.CoveredCount))
	}
	return strings.Join(lines, "\n") + "\n"
}

func RunReport(relativeFile string, baseline mutruntime.RunResult, diagnostics string, uncovered []analysis.MutationSite, results []mutruntime.JobResult) string {
	lines := []string{fmt.Sprintf("Baseline tests passed in %d ms.", baseline.DurationMS), strings.TrimSuffix(diagnostics, "\n")}
	for _, site := range uncovered {
		lines = append(lines, fmt.Sprintf("UNCOVERED %s:%d %s", relativeFile, site.Line, site.Description))
	}
	killed := 0
	for _, result := range results {
		if result.Killed {
			killed++
		}
		status := "SURVIVED"
		if result.Killed {
			status = "KILLED"
		}
		lines = append(lines, fmt.Sprintf("%s %s:%d %s (%d ms, %d ms wall)", status, relativeFile, result.Line, result.Description, result.DurationMS, result.JobWallMS))
		if result.TimedOut {
			lines = append(lines, "  timed out")
		}
	}
	lines = append(lines, fmt.Sprintf("Coverage: %d uncovered sites skipped.", len(uncovered)))
	lines = append(lines, fmt.Sprintf("Summary: %d killed, %d survived, %d total.", killed, len(results)-killed, len(results)))
	return strings.Join(lines, "\n") + "\n"
}
